File: tools/lint_check.py
#!/usr/bin/env python3
from __future__ import annotations

import argparse
import json
import re
import shutil
import subprocess
import sys
from dataclasses import dataclass
from pathlib import Path
from re import Pattern
from typing import Protocol

from lint import architecture, include_style, source_dependencies, source_policy
from lint.common import (
    CheckResult,
    CheckerContext,
    Config,
    FileRecord,
    IncludeDirective,
    config_strings,
    format_count,
    has_root,
    is_excluded,
    project_path,
)


PROJECT_ROOT = Path(__file__).resolve().parents[1]
CONFIG_PATH = Path(__file__).with_name("lint_config.json")


class Checker(Protocol):
    def process_file(self, record: FileRecord) -> None:
        ...

    def finish(self, verbose: bool) -> CheckResult:
        ...


@dataclass(frozen=True)
class FileEntry:
    path: Path
    tracked: bool


@dataclass(frozen=True)
class ScanSettings:
    roots: tuple[str, ...]
    suffixes: frozenset[str]
    source_suffixes: frozenset[str]
    tracked_only_suffixes: frozenset[str]
    excluded_prefixes: tuple[str, ...]
    include_pattern: Pattern[str]
    nolint_pattern: Pattern[str]
    wide_literal_allowance: Pattern[str]


def relpath(path: Path) -> str:
    return path.relative_to(PROJECT_ROOT).as_posix()


def load_config() -> Config:
    return json.loads(CONFIG_PATH.read_text(encoding="utf-8"))


def parse_scan_settings(config: Config) -> ScanSettings:
    scan_config = config["scan"]
    return ScanSettings(
        roots=config_strings(scan_config, "roots"),
        suffixes=frozenset(config_strings(scan_config, "suffixes")),
        source_suffixes=frozenset(config_strings(scan_config, "source_suffixes")),
        tracked_only_suffixes=frozenset(config_strings(scan_config, "tracked_only_suffixes")),
        excluded_prefixes=config_strings(scan_config, "excluded_prefixes"),
        include_pattern=re.compile(str(scan_config["include_pattern"])),
        nolint_pattern=re.compile(str(scan_config["nolint_pattern"])),
        wide_literal_allowance=source_policy.compile_wide_literal_allowance(config["source_policy"]),
    )


def run_git_ls_files(args: list[str]) -> list[str] | None:
    try:
        result = subprocess.run(
            ["git", "ls-files", *args],
            cwd=PROJECT_ROOT,
            check=True,
            capture_output=True,
            text=True,
        )
    except (OSError, subprocess.CalledProcessError):
        return None
    return [line.strip() for line in result.stdout.splitlines() if line.strip()]


def project_paths_from_git(lines: list[str]) -> set[Path]:
    return {(PROJECT_ROOT / line).resolve() for line in lines}


def is_lint_input(path: Path, tracked: bool, settings: ScanSettings) -> bool:
    relative = relpath(path)
    if path.suffix not in settings.suffixes:
        return False
    if settings.roots and not has_root(relative, settings.roots):
        return False
    if is_excluded(relative, settings.excluded_prefixes):
        return False
    if path.suffix in settings.tracked_only_suffixes and not tracked:
        return False
    return True


def discover_lint_inputs(settings: ScanSettings) -> list[FileEntry]:
    tracked_lines = run_git_ls_files(list(settings.roots))
    untracked_lines = run_git_ls_files(["--others", "--exclude-standard", *settings.roots])

    if tracked_lines is None or untracked_lines is None:
        paths = {
            path.resolve()
            for root in settings.roots
            for path in (PROJECT_ROOT / root).rglob("*")
            if path.is_file()
        }
        return [
            FileEntry(path=path, tracked=True)
            for path in sorted(paths, key=relpath)
            if is_lint_input(path, tracked=True, settings=settings)
        ]

    tracked_paths = project_paths_from_git(tracked_lines)
    untracked_paths = project_paths_from_git(untracked_lines)
    all_paths = sorted(tracked_paths | untracked_paths, key=relpath)
    return [
        FileEntry(path=path, tracked=path in tracked_paths)
        for path in all_paths
        if path.is_file() and is_lint_input(path, tracked=path in tracked_paths, settings=settings)
    ]


def scan_file(entry: FileEntry, settings: ScanSettings) -> FileRecord:
    text = entry.path.read_text(encoding="utf-8")
    lines = tuple(text.splitlines())
    includes: list[IncludeDirective] = []
    nolint_lines: list[int] = []

    for line_number, line in enumerate(lines, start=1):
        if settings.nolint_pattern.search(line):
            nolint_lines.append(line_number)

        match = settings.include_pattern.match(line)
        if match:
            quoted_text = match.group(1)
            includes.append(
                IncludeDirective(
                    line=line_number,
                    text=quoted_text or match.group(2),
                    quoted=quoted_text is not None,
                )
            )

    stripped_text = ""
    stripped_lines: tuple[str, ...] = ()
    wide_literal_lines: tuple[int, ...] = ()
    if entry.path.suffix in settings.source_suffixes:
        stripped_text = source_policy.strip_comments_and_strings(text)
        stripped_lines = tuple(stripped_text.splitlines())
        wide_literal_lines = tuple(source_policy.find_undocumented_wide_literal_lines(text, settings.wide_literal_allowance))

    return FileRecord(
        path=entry.path,
        relative=relpath(entry.path),
        tracked=entry.tracked,
        text=text,
        lines=lines,
        includes=tuple(includes),
        nolint_lines=tuple(nolint_lines),
        stripped_text=stripped_text,
        stripped_lines=stripped_lines,
        wide_literal_lines=wide_literal_lines,
    )


def truncate_progress_line(prefix: str, relative: str) -> str:
    columns = shutil.get_terminal_size((120, 20)).columns
    if columns <= 1:
        return f"{prefix}{relative}"
    max_length = columns - 1
    full_line = f"{prefix}{relative}"
    if len(full_line) <= max_length:
        return full_line
    path_budget = max_length - len(prefix)
    if path_budget <= 3:
        return full_line[:max_length]
    return f"{prefix}...{relative[-(path_budget - 3):]}"


def scan_lint_inputs(entries: list[FileEntry], settings: ScanSettings, checkers: tuple[Checker, ...], show_progress: bool) -> tuple[FileRecord, ...]:
    records: list[FileRecord] = []
    total = len(entries)
    use_single_line_progress = show_progress and sys.stdout.isatty()
    previous_progress_length = 0
    for index, entry in enumerate(entries, start=1):
        if use_single_line_progress:
            relative = relpath(entry.path)
            progress = truncate_progress_line(f"[{index}/{total}] lint-check ", relative)
            padding = " " * max(0, previous_progress_length - len(progress))
            sys.stdout.write(f"\r{progress}{padding}")
            sys.stdout.flush()
            previous_progress_length = len(progress)
        record = scan_file(entry, settings)
        for checker in checkers:
            checker.process_file(record)
        records.append(record)
    if use_single_line_progress:
        sys.stdout.write(f"\r{' ' * previous_progress_length}\r")
        sys.stdout.flush()
    return tuple(records)


def absolute_project_path(path: Path) -> Path:
    return path if path.is_absolute() else PROJECT_ROOT / path


def parse_args(config: Config) -> argparse.Namespace:
    source_dependency_config = config["source_dependencies"]
    default_dot_path = project_path(PROJECT_ROOT, str(source_dependency_config["default_dot_output"]))
    parser = argparse.ArgumentParser(
        description="Run combined architecture, source dependency, include style, and source policy lint checks."
    )
    parser.add_argument(
        "--output",
        "-o",
        type=Path,
        default=default_dot_path,
        help=f"DOT output path. Defaults to {default_dot_path.relative_to(PROJECT_ROOT)}.",
    )
    parser.add_argument(
        "--graphml-output",
        type=Path,
        default=None,
        help="GraphML output path. Defaults to the DOT output path with a .graphml suffix.",
    )
    parser.add_argument(
        "--svg-output",
        type=Path,
        default=None,
        help="SVG output path. Defaults to the DOT output path with a .svg suffix.",
    )
    parser.add_argument(
        "--skip-svg",
        action="store_true",
        help="Skip Graphviz SVG rendering and write only DOT and GraphML outputs.",
    )
    parser.add_argument(
        "--check",
        action="store_true",
        help="Fail when source dependency architecture rules are violated.",
    )
    parser.add_argument(
        "--no-progress",
        action="store_true",
        help="Do not print per-file scan progress.",
    )
    parser.add_argument(
        "-v",
        "--verbose",
        action="store_true",
        help="Print source files over the configured LOC threshold and topological dependency details.",
    )
    return parser.parse_args()


def create_context(args: argparse.Namespace, settings: ScanSettings) -> CheckerContext:
    dot_output = absolute_project_path(args.output)
    graphml_output = absolute_project_path(args.graphml_output) if args.graphml_output else dot_output.with_suffix(".graphml")
    svg_output = absolute_project_path(args.svg_output) if args.svg_output else dot_output.with_suffix(".svg")
    return CheckerContext(
        project_root=PROJECT_ROOT,
        excluded_prefixes=settings.excluded_prefixes,
        check_dependencies=bool(args.check),
        dot_output=dot_output,
        graphml_output=graphml_output,
        svg_output=svg_output,
        skip_svg=bool(args.skip_svg),
    )


def create_checkers(config: Config, context: CheckerContext) -> tuple[Checker, ...]:
    return (
        architecture.create_checker(config["architecture"], context),
        source_dependencies.create_checker(config["source_dependencies"], context),
        include_style.create_checker(config["include_style"], context),
        source_policy.create_checker(config["source_policy"], context),
    )


def print_scanned_loc_summary(records: tuple[FileRecord, ...]) -> None:
    total_loc = sum(record.line_count for record in records)
    print(f"Scanned {format_count(total_loc)} LOC across {len(records)} lint input file(s).")


def print_failure_result(result: CheckResult) -> None:
    print(result.title)
    for error in result.errors:
        print(error)
    for finding in result.findings:
        print(f"{finding.location}: {finding.kind}: {finding.message}")
    if result.summary:
        print(result.summary)


def main() -> int:
    config = load_config()
    settings = parse_scan_settings(config)
    args = parse_args(config)
    context = create_context(args, settings)
    checkers = create_checkers(config, context)

    records = scan_lint_inputs(
        discover_lint_inputs(settings),
        settings,
        checkers,
        show_progress=not args.no_progress,
    )
    results = tuple(checker.finish(args.verbose) for checker in checkers)

    printed_report = False
    for result in results:
        if not result.failed:
            continue
        if printed_report:
            print()
        print_failure_result(result)
        printed_report = True

    if any(result.failed for result in results):
        if printed_report:
            print()
        print("Combined lint check failed.")
        return 1

    print_scanned_loc_summary(records)
    if args.verbose:
        verbose_lines = [line for result in results for line in result.verbose_lines]
        if verbose_lines:
            print()
            for line in verbose_lines:
                print(line)
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: src/tools/lint_check.cpp
#include "tools/lint_check.h"

#include <algorithm>
#include <chrono>
#include <cstdio>
#include <regex>
#include <set>
#include <stdexcept>

#include "tools/impl/lint_checkers.h"
#include "tools/impl/lint_common.h"
#include "tools/impl/lint_json.h"
#include "tools/impl/tools_common.h"
#include "tools/impl/tools_parallel.h"
#include "tools/impl/tools_progress.h"
#include "util/file_path.h"
#include "util/strings.h"

namespace tools::lint {

namespace {

constexpr int kToolStaleExitCode = 3;

const std::vector<std::string>& ToolRefreshExcludedPrefixes() {
    static const std::vector<std::string> prefixes = {"src/vendor/", "src/tools/vendor/"};
    return prefixes;
}

struct LintArgs {
    std::string configPath;
    bool check = false;
    bool noProgress = false;
    bool verbose = false;
    size_t concurrency = 0;
    std::optional<std::string> reportJson;
};

struct CompletedLintScan {
    FileRecord record;
    std::string error;
};

bool IsNewerThan(std::string_view source, std::uint64_t targetTime) {
    const std::optional<std::uint64_t> sourceTime = LastWriteTime(source);
    return sourceTime.has_value() && *sourceTime > targetTime;
}

bool IsCMakeBuildGraphStale(const std::string& repoRoot) {
    const std::optional<std::uint64_t> cmakeListsTime = LastWriteTime((FilePath(repoRoot) / "CMakeLists.txt").string());
    if (!cmakeListsTime.has_value()) {
        return false;
    }

    // A CMake edit can refresh build.ninja without relinking CaseDashTools when the tool target is unchanged.
    const std::optional<std::uint64_t> buildGraphTime =
        LastWriteTime((FilePath(repoRoot) / "build/cmake/build.ninja").string());
    return !buildGraphTime.has_value() || *cmakeListsTime > *buildGraphTime;
}

bool IsToolStale() {
    const std::string exePath = ExecutablePath();
    const std::optional<std::uint64_t> exeTime = LastWriteTime(exePath);
    if (!exeTime.has_value()) {
        return false;
    }
    const std::string repoRoot = FilePath(FilePath(exePath).ParentPath().string()).ParentPath().string();
    if (IsCMakeBuildGraphStale(repoRoot)) {
        return true;
    }
    for (const std::string& path : RecursiveFiles((FilePath(repoRoot) / "src/tools").string())) {
        const std::string relative = RelativePath(path, repoRoot);
        // The freshness check runs before lint config parsing, so keep vendored roots out here as well.
        if (IsExcluded(relative, ToolRefreshExcludedPrefixes())) {
            continue;
        }
        const std::string suffix = Extension(path);
        if ((suffix == ".cpp" || suffix == ".h") && IsNewerThan(path, *exeTime)) {
            return true;
        }
    }
    return false;
}

std::string ResolveProjectPath(const std::string& projectRoot, const std::string& path) {
    const bool isAbsolute =
        (path.size() >= 2 && path[1] == ':') || (!path.empty() && (path[0] == '/' || path[0] == '\\'));
    if (isAbsolute) {
        return AbsolutePath(path);
    }
    return AbsolutePath((FilePath(projectRoot) / path).string());
}

LintArgs ParseArgs(int argc, char** argv, const std::string& projectRoot) {
    LintArgs args;
    args.configPath = (FilePath(projectRoot) / "tools/lint_config.json").string();
    for (int i = 0; i < argc; ++i) {
        const std::string arg = argv[i];
        if (arg == "--config") {
            if (i + 1 >= argc) {
                throw std::runtime_error("--config requires a value");
            }
            args.configPath = ResolveProjectPath(projectRoot, argv[++i]);
        } else if (arg == "--check") {
            args.check = true;
        } else if (arg == "--no-progress") {
            args.noProgress = true;
        } else if (arg == "--concurrency") {
            if (i + 1 >= argc) {
                throw std::runtime_error("--concurrency requires a value");
            }
            std::string error;
            if (!ParseToolConcurrency(argv[++i], args.concurrency, error)) {
                throw std::runtime_error(error);
            }
        } else if (StartsWith(arg, "--concurrency=")) {
            std::string error;
            if (!ParseToolConcurrency(std::string_view(arg).substr(14), args.concurrency, error)) {
                throw std::runtime_error(error);
            }
        } else if (arg == "--report-json") {
            if (i + 1 >= argc) {
                throw std::runtime_error("--report-json requires a value");
            }
            args.reportJson = argv[++i];
        } else if (arg == "-v" || arg == "--verbose") {
            args.verbose = true;
        } else {
            throw std::runtime_error("unknown argument " + arg);
        }
    }
    args.configPath = ResolveProjectPath(projectRoot, args.configPath);
    return args;
}

std::map<std::string, std::set<std::string>> ParseSuffixGroups(const JsonValue& config) {
    std::map<std::string, std::set<std::string>> groups;
    for (const auto& [name, values] : config.At("suffix_groups").AsObject()) {
        std::set<std::string> suffixes;
        for (const JsonValue& value : values.AsArray()) {
            suffixes.insert(value.AsString());
        }
        groups[name] = std::move(suffixes);
    }
    return groups;
}

void RequireSingleSuffixGroup(
    const std::map<std::string, std::set<std::string>>& suffixGroups,
    std::string_view configPath,
    std::string_view groupName
) {
    const std::set<std::string> values = RequireSuffixGroup(suffixGroups, configPath, groupName);
    if (values.size() != 1) {
        throw std::runtime_error(std::string(configPath) + " must reference a suffix group with exactly one suffix");
    }
}

void ValidateConfig(const JsonValue& config, const std::map<std::string, std::set<std::string>>& suffixGroups) {
    RequireSuffixGroup(suffixGroups, "scan.suffix_group", config.At("scan").At("suffix_group").AsString());
    RequireSingleSuffixGroup(
        suffixGroups,
        "architecture.header_suffix_group",
        config.At("architecture").At("header_suffix_group").AsString()
    );
    RequireSingleSuffixGroup(
        suffixGroups,
        "architecture.implementation_suffix_group",
        config.At("architecture").At("implementation_suffix_group").AsString()
    );
    RequireSuffixGroup(
        suffixGroups,
        "include_style.suffix_group",
        config.At("include_style").At("suffix_group").AsString()
    );
    RequireSuffixGroup(
        suffixGroups,
        "source_dependencies.suffix_group",
        config.At("source_dependencies").At("suffix_group").AsString()
    );
    RequireSuffixGroup(
        suffixGroups,
        "source_dependencies.header_suffix_group",
        config.At("source_dependencies").At("header_suffix_group").AsString()
    );
    RequireSuffixGroup(
        suffixGroups,
        "source_policy.suffix_group",
        config.At("source_policy").At("suffix_group").AsString()
    );
    if (ConfigStrings(config.At("source_dependencies"), "roots").size() != 1) {
        throw std::runtime_error("source_dependencies.roots must contain exactly one root");
    }
}

ScanSettings
    ParseScanSettings(const JsonValue& config, const std::map<std::string, std::set<std::string>>& suffixGroups)
{
    const JsonValue& scan = config.At("scan");
    ScanSettings settings;
    settings.roots = ConfigStrings(scan, "roots");
    settings.suffixes = RequireSuffixGroup(suffixGroups, "scan.suffix_group", scan.At("suffix_group").AsString());
    settings.strippedSuffixes = RequireSuffixGroup(suffixGroups, "suffix_groups.source", "source");
    settings.excludedPrefixes = ConfigStrings(scan, "excluded_prefixes");
    settings.includePattern = scan.At("include_pattern").AsString();
    return settings;
}

bool IsLintInput(const FileEntry& entry, const std::string& projectRoot, const ScanSettings& settings) {
    const std::string relative = RelativePath(entry.path, projectRoot);
    if (settings.suffixes.find(Extension(entry.path)) == settings.suffixes.end()) {
        return false;
    }
    if (!settings.roots.empty() && !HasRoot(relative, settings.roots)) {
        return false;
    }
    return !IsExcluded(relative, settings.excludedPrefixes);
}

std::set<std::string> ProjectPathsFromGit(const std::string& projectRoot, const std::vector<std::string>& lines) {
    std::set<std::string> paths;
    for (const std::string& line : lines) {
        paths.insert(NormalizePathKey((FilePath(projectRoot) / line).string()));
    }
    return paths;
}

std::vector<FileEntry> DiscoverLintInputs(const std::string& projectRoot, const ScanSettings& settings) {
    const std::optional<std::vector<std::string>> trackedLines = RunGitLsFiles(settings.roots);
    std::vector<std::string> untrackedArgs = {"--others", "--exclude-standard"};
    untrackedArgs.insert(untrackedArgs.end(), settings.roots.begin(), settings.roots.end());
    const std::optional<std::vector<std::string>> untrackedLines = RunGitLsFiles(untrackedArgs);

    std::vector<FileEntry> entries;
    if (!trackedLines.has_value() || !untrackedLines.has_value()) {
        for (const std::string& root : settings.roots) {
            for (const std::string& path : RecursiveFiles((FilePath(projectRoot) / root).string())) {
                FileEntry entry{path, true};
                if (IsLintInput(entry, projectRoot, settings)) {
                    entries.push_back(std::move(entry));
                }
            }
        }
    } else {
        const std::set<std::string> trackedPaths = ProjectPathsFromGit(projectRoot, *trackedLines);
        std::set<std::string> allPaths = trackedPaths;
        const std::set<std::string> untrackedPaths = ProjectPathsFromGit(projectRoot, *untrackedLines);
        allPaths.insert(untrackedPaths.begin(), untrackedPaths.end());
        for (const std::string& key : allPaths) {
            FileEntry entry{AbsolutePath(key), trackedPaths.find(key) != trackedPaths.end()};
            if (FileExists(FilePath(entry.path)) && IsLintInput(entry, projectRoot, settings)) {
                entries.push_back(std::move(entry));
            }
        }
    }

    std::sort(entries.begin(), entries.end(), [&](const FileEntry& left, const FileEntry& right) {
        return RelativePath(left.path, projectRoot) < RelativePath(right.path, projectRoot);
    });
    return entries;
}

FileRecord ScanFile(const FileEntry& entry, const std::string& projectRoot, const ScanSettings& settings) {
    static const std::regex includePattern(R"include(^\s*#include\s+("([^"]+)"|<([^>]+)>))include");

    const std::optional<std::string> text = ReadFileBinary(entry.path);
    if (!text.has_value()) {
        throw std::runtime_error("could not read " + entry.path);
    }

    FileRecord record;
    record.path = AbsolutePath(entry.path);
    record.relative = RelativePath(record.path, projectRoot);
    record.tracked = entry.tracked;
    record.text = *text;
    record.lines = SplitLines(record.text);

    for (int index = 0; index < static_cast<int>(record.lines.size()); ++index) {
        const std::string& line = record.lines[static_cast<size_t>(index)];
        std::smatch match;
        if (!std::regex_search(line, match, includePattern)) {
            continue;
        }
        const bool quoted = match.length(2) > 0;
        record.includes.push_back({index + 1, quoted ? match.str(2) : match.str(3), quoted});
    }

    if (settings.strippedSuffixes.find(Extension(entry.path)) != settings.strippedSuffixes.end()) {
        record.strippedText = StripCommentsAndStrings(record.text);
        record.strippedLines = SplitLines(record.strippedText);
    }
    return record;
}

std::vector<FileRecord> ScanLintInputs(
    const std::vector<FileEntry>& entries,
    const std::string& projectRoot,
    const ScanSettings& settings,
    size_t concurrency,
    bool showProgress,
    std::chrono::steady_clock::time_point started
) {
    std::vector<CompletedLintScan> completed(entries.size());
    ToolFileProgress progress(stdout, "lint-check", entries.size(), started, showProgress);
    RunToolParallelFor(entries.size(), concurrency, &progress, [&](size_t index) {
        try {
            completed[index].record = ScanFile(entries[index], projectRoot, settings);
        } catch (const std::exception& error) {
            completed[index].error = error.what();
        }
    });

    std::vector<FileRecord> records;
    records.reserve(entries.size());
    for (CompletedLintScan& scan : completed) {
        if (!scan.error.empty()) {
            throw std::runtime_error(scan.error);
        }
        records.push_back(std::move(scan.record));
    }
    return records;
}

std::string CheckName(const CheckResult& result) {
    std::string title = result.title;
    if (EndsWith(title, ":")) {
        title.pop_back();
    }
    if (EndsWith(title, " check")) {
        title.resize(title.size() - 6);
    }
    title = ToLower(Trim(title));
    std::replace(title.begin(), title.end(), ' ', '_');
    return title;
}

std::vector<Diagnostic> DiagnosticsForResult(const CheckResult& result) {
    std::vector<Diagnostic> diagnostics;
    const std::string check = CheckName(result);
    for (const std::string& error : result.errors) {
        diagnostics.push_back({check, "error", {}, {}, error});
    }
    for (const Finding& finding : result.findings) {
        diagnostics.push_back({check, "finding", finding.location, finding.kind, finding.message});
    }
    return diagnostics;
}

std::vector<Diagnostic> CollectDiagnostics(const std::vector<CheckResult>& results) {
    std::vector<Diagnostic> diagnostics;
    for (const CheckResult& result : results) {
        const std::vector<Diagnostic> resultDiagnostics = DiagnosticsForResult(result);
        diagnostics.insert(diagnostics.end(), resultDiagnostics.begin(), resultDiagnostics.end());
    }
    return diagnostics;
}

bool WriteReportJson(const std::string& path, bool failed, const std::vector<Diagnostic>& diagnostics) {
    std::string json;
    json += "{\n";
    json += "  \"schema_version\": 1,\n";
    json += std::string("  \"failed\": ") + (failed ? "true" : "false") + ",\n";
    json += "  \"diagnostics\": [\n";
    for (size_t i = 0; i < diagnostics.size(); ++i) {
        const Diagnostic& diagnostic = diagnostics[i];
        json += "    {\n";
        json += "      \"check\": \"" + JsonEscape(diagnostic.check) + "\",\n";
        json += "      \"type\": \"" + JsonEscape(diagnostic.type) + "\",\n";
        if (!diagnostic.location.empty()) {
            json += "      \"location\": \"" + JsonEscape(diagnostic.location) + "\",\n";
            json += "      \"kind\": \"" + JsonEscape(diagnostic.kind) + "\",\n";
        }
        json += "      \"message\": \"" + JsonEscape(diagnostic.message) + "\"\n";
        json += "    }";
        json += i + 1 == diagnostics.size() ? "\n" : ",\n";
    }
    json += "  ]\n";
    json += "}\n";
    return EnsureParentDirectory(path) && WriteFileBinary(path, json);
}

std::string ScannedFileCountText(size_t scannedFiles, size_t totalFiles) {
    if (scannedFiles == totalFiles) {
        return std::to_string(scannedFiles);
    }
    return std::to_string(scannedFiles) + "/" + std::to_string(totalFiles);
}

void PrintScannedLocSummary(const std::vector<FileRecord>& records, size_t totalFiles) {
    int loc = 0;
    for (const FileRecord& record : records) {
        loc += record.LineCount();
    }
    const std::string scannedFiles = ScannedFileCountText(records.size(), totalFiles);
    std::printf("Scanned %s LOC across %s lint input file(s).\n", FormatCount(loc).c_str(), scannedFiles.c_str());
}

void PrintFailureResult(const CheckResult& result) {
    std::printf("%s\n", result.title.c_str());
    for (const Diagnostic& diagnostic : DiagnosticsForResult(result)) {
        if (diagnostic.type == "error") {
            std::printf("%s\n", diagnostic.message.c_str());
        } else {
            std::printf(
                "%s: %s: %s\n",
                diagnostic.location.c_str(),
                diagnostic.kind.c_str(),
                diagnostic.message.c_str()
            );
        }
    }
    if (!result.summary.empty()) {
        std::printf("%s\n", result.summary.c_str());
    }
}

}  // namespace

}  // namespace tools::lint

int RunLintCheck(int argc, char** argv) {
    using namespace tools::lint;

    const auto started = std::chrono::steady_clock::now();
    if (IsToolStale()) {
        std::fprintf(stderr, "CaseDashTools build inputs changed; rebuilding.\n");
        return kToolStaleExitCode;
    }

    const std::string projectRoot = AbsolutePath(CurrentDirectoryPath().string());
    LintArgs args;
    JsonValue config;
    std::map<std::string, std::set<std::string>> suffixGroups;
    ScanSettings settings;
    try {
        args = ParseArgs(argc, argv, projectRoot);
        const std::optional<std::string> configText = ReadFileBinary(args.configPath);
        if (!configText.has_value()) {
            throw std::runtime_error("could not read " + args.configPath);
        }
        config = ParseJson(*configText);
        suffixGroups = ParseSuffixGroups(config);
        ValidateConfig(config, suffixGroups);
        settings = ParseScanSettings(config, suffixGroups);
    } catch (const std::exception& error) {
        std::fprintf(stderr, "lint config error: %s\n", error.what());
        std::printf("Lint failed in %s.\n", FormatToolElapsed(std::chrono::steady_clock::now() - started).c_str());
        return 2;
    }

    CheckerContext context;
    context.projectRoot = projectRoot;
    context.suffixGroups = suffixGroups;
    context.excludedPrefixes = settings.excludedPrefixes;
    context.checkDependencies = args.check;

    std::vector<std::unique_ptr<Checker>> checkers;
    try {
        checkers = CreateCheckers(config, context);
    } catch (const std::exception& error) {
        std::fprintf(stderr, "lint config error: %s\n", error.what());
        std::printf("Lint failed in %s.\n", FormatToolElapsed(std::chrono::steady_clock::now() - started).c_str());
        return 2;
    }

    std::vector<FileEntry> entries;
    std::vector<FileRecord> records;
    std::vector<CheckResult> results;
    try {
        entries = DiscoverLintInputs(projectRoot, settings);
        records = ScanLintInputs(entries, projectRoot, settings, args.concurrency, !args.noProgress, started);
        for (const FileRecord& record : records) {
            for (std::unique_ptr<Checker>& checker : checkers) {
                try {
                    checker->ProcessFile(record);
                } catch (const std::exception& error) {
                    throw std::runtime_error("while processing " + record.relative + ": " + error.what());
                }
            }
        }
        for (std::unique_ptr<Checker>& checker : checkers) {
            results.push_back(checker->Finish(args.verbose));
        }
    } catch (const std::exception& error) {
        std::fprintf(stderr, "lint error: %s\n", error.what());
        const std::string scannedFiles = ScannedFileCountText(records.size(), entries.size());
        std::printf(
            "Lint failed after scanning %s file(s) in %s.\n",
            scannedFiles.c_str(),
            FormatToolElapsed(std::chrono::steady_clock::now() - started).c_str()
        );
        return 2;
    }

    const bool failed =
        std::any_of(results.begin(), results.end(), [](const CheckResult& result) { return result.Failed(); });
    const std::vector<Diagnostic> diagnostics = CollectDiagnostics(results);

    if (args.reportJson.has_value()) {
        const std::string reportPath = ResolveProjectPath(projectRoot, *args.reportJson);
        if (!WriteReportJson(reportPath, failed, diagnostics)) {
            std::fprintf(stderr, "lint report error: could not write %s\n", reportPath.c_str());
            const std::string scannedFiles = ScannedFileCountText(records.size(), entries.size());
            std::printf(
                "Lint failed after scanning %s file(s) in %s.\n",
                scannedFiles.c_str(),
                FormatToolElapsed(std::chrono::steady_clock::now() - started).c_str()
            );
            return 2;
        }
    }

    bool printedReport = false;
    for (const CheckResult& result : results) {
        if (!result.Failed()) {
            continue;
        }
        if (printedReport) {
            std::printf("\n");
        }
        PrintFailureResult(result);
        printedReport = true;
    }

    if (failed) {
        if (printedReport) {
            std::printf("\n");
        }
        const std::string scannedFiles = ScannedFileCountText(records.size(), entries.size());
        std::printf(
            "Lint failed after scanning %s file(s) in %s.\n",
            scannedFiles.c_str(),
            FormatToolElapsed(std::chrono::steady_clock::now() - started).c_str()
        );
        return 1;
    }

    PrintScannedLocSummary(records, entries.size());
    if (args.verbose) {
        std::vector<std::string> verboseLines;
        for (const CheckResult& result : results) {
            verboseLines.insert(verboseLines.end(), result.verboseLines.begin(), result.verboseLines.end());
        }
        if (!verboseLines.empty()) {
            std::printf("\n");
            for (const std::string& line : verboseLines) {
                std::printf("%s\n", line.c_str());
            }
        }
    }
    std::printf(
        "Lint succeeded after scanning %s file(s) in %s.\n",
        ScannedFileCountText(records.size(), entries.size()).c_str(),
        FormatToolElapsed(std::chrono::steady_clock::now() - started).c_str()
    );
    return 0;
}
